src/sync.rs

```diff
@@ -527,6 +531,100 @@ where
     }
 }
 
+/// Executes a closure with a Python critical section held on mutex.
+///
+/// Acquires the per-object lock for the mutex `mutex` that is held
+/// until the closure `f` is finished.
+///
+/// This is structurally equivalent to the use of the paired
+/// Py_BEGIN_CRITICAL_SECTION_MUTEX and Py_END_CRITICAL_SECTION C-API macros.
+///
+/// A no-op on GIL-enabled builds, where the critical section API is exposed as
+/// a no-op by the Python C API.
+///
+/// Provides weaker locking guarantees than traditional locks, but can in some
+/// cases be used to provide guarantees similar to the GIL without the risk of
+/// deadlocks associated with traditional locks.
+///
+/// This variant with a mutex as an argument is particularly useful when paired
+/// with a static `PyMutex`. This can be used in situations where a `PyMutex` would
+/// deadlock with the interpreter.
+///
+/// Many CPython C API functions do not acquire the per-object lock on objects
+/// passed to Python. You should not expect critical sections applied to
+/// built-in types to prevent concurrent modification. This API is most useful
+/// for user-defined types with full control over how the internal state for the
+/// type is managed.
+///
+/// Only available on Python 3.14 and newer.
+#[cfg(Py_3_14)]
+#[cfg_attr(not(Py_GIL_DISABLED), allow(unused_variables))]
+pub fn with_critical_section_mutex<F, R, T>(mutex: &mut PyMutex<T>, f: F) -> R
+where
+    F: FnOnce(&mut T) -> R,
+{
+    #[cfg(Py_GIL_DISABLED)]
+    {
+        let mut guard = CSGuard(unsafe { std::mem::zeroed() });
+        unsafe { crate::ffi::PyCriticalSection_BeginMutex(&mut guard.0, &mut *mutex.mutex.get()) };
+        f(mutex.data.get_mut())
```

Comment:
I guess we might also need to worry about panicking here? e.g. if we panic within this closure, we should probably poison the mutex?